californium-core/src/main/java/org/eclipse/californium/core/network/Exchange.java

```diff
@@ -201,6 +198,19 @@
 	 * Mark exchange as notification.
 	 */
 	private final boolean notification;
+	/**
+	 * The key mid for the current request.
+	 */
+	private KeyMID currentKeyMID;
+	/**
+	 * The key token for the original request, if {@link #keepRequestInStore} is
+	 * {@code true}. {@code null} otherwise.
+	 */
+	private KeyToken originalKeyToken;
+	/**
+	 * The key token for the current request.
+	 */
+	private KeyToken currentKeyToken;
```

Comment:
The new idea is to save the keys after successfully put in the exchange itself. That simplifies the house keeping.